### PR TITLE
FEATURE: Assets Eel helper for Neos.Media

### PR DESCRIPTION
**Pull request by Sebobo**
This helper provides access to assets in Neos by either searching by Tag, AssetCollection, or
a searchTerm combined with optionally Tags and an AssetCollection.

**Review instructions**

Add some assets to collections and tag them, then use the helpers in Fusion like this:

```
assetsByTag = ${Neos.Media.Assets.findByTag('MyTag')}
assetsByCollection = ${Neos.Media.Assets.findByCollection('MyCollection')}
assetsBySearchTerm = ${Neos.Media.Assets.findBySearchTerm('alice')}
```

**Checklist**

- [x] Code follows the PSR-2 coding style
- [x] Tests have been created, run and adjusted as needed
- [x] The PR is created against the [lowest maintained branch](https://www.neos.io/features/release-roadmap.html)
- [x] Reviewer - PR Title is brief but complete and starts with `FEATURE|TASK|BUGFIX`
- [x] Reviewer - The first section explains the change briefly for change-logs
- [ ] Reviewer - Breaking Changes are marked with `!!!` and have upgrade-instructions

### CODE REVIEW

**Comment by Sebobo** [comment]
I thought also about adding datasource to allow the selection of tags and collections, but that would require adding Neos as dependency to Neos.Media, which is Quatsch.

But it would help a lot with making it possible to select Tags and Collections as node property combined with this helper.
So an alternative would be to add them to Neos.Neos instead or to Flowpack.MediaUI 🤔 

And currently the `searchTerm` method is not including the optional tags and collection in its name. But somehow `findBySearchTermAndTagsAndCollections` is also silly...

File: Neos.Media/Classes/Eel/AssetsHelper.php [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+            $tag = $this->tagRepository->findOneByLabel($tag);
+        }
+
+        if (!$tag) {
+            return null;
+        }
+
+        try {
+            return $this->assetRepository->findByTag($tag);
+        } catch (InvalidQueryException) {
```
Comment (mhsdesign) [inline Neos.Media/Classes/Eel/AssetsHelper.php]: `InvalidQueryException` is actually never at any place thrown 😂 

lets see ... https://neos-project.slack.com/archives/C050KKBEB/p1680344642064619

**Comment by kitsunet** [comment]
I would suggest creating separete methods for the respective ObjectType|string inputs, eg. `findByLabel` and `findByLabelName` or something, same for the other. Makes for a cleaner interface?

**Comment by mhsdesign** [comment]
> I would suggest creating separete methods for the respective ObjectType|string inputs, eg. `findByLabel` and `findByLabelName` or something, same for the other. Makes for a cleaner interface?


i think we should follow the mentality of the new eel helpers for the ESCR ... idont know exactly how they look like, but they must have the same challenge of dtos vs primitives ...

File: Neos.Media/Classes/Eel/AssetsHelper.php (hunk 1 above)
Comment (Sebobo) [reply Neos.Media/Classes/Eel/AssetsHelper.php]: PHPStorm was certain about it 😉 but can check if I can remove the annotation and then the try block

**Comment by kitsunet** [comment]
> > I would suggest creating separete methods for the respective ObjectType|string inputs, eg. `findByLabel` and `findByLabelName` or something, same for the other. Makes for a cleaner interface?
> 
> i think we should follow the mentality of the new eel helpers for the ESCR ... idont know exactly how they look like, but they must have the same challenge of dtos vs primitives ...

I would definitely hard suggest it there too 😂

**Comment by Sebobo** [comment]
And do you guys think I should add any other methods from the AssetRepo?

**Comment by bwaidelich** [comment]
> i think we should follow the mentality of the new eel helpers for the ESCR

There is not really a (new) mentality as far as I know. But I would suggest to see Eel helpers as adapters from the loosely typed fusion world into a strictly typed PHP world and therefore keep union types where it makes sense (i.e. where multiple variants *mean the same thing*).
e.g. I found it rather confusing to have `findByLabel` and `findByLabelName` on a fusion level.

But I think @kitsunet's comment was re `findByTag` and I agree that `findByTag` and `findByTagLabel` are two different things.

More concretely:

I think this is totally fine and makes it easier to use a helper:

```php
function someHelper(SomeValueObject|string|null $foo): ?string {
    if ($foo === null) {
        return null;
    }
    if (is_string($foo)) {
       $foo = SomeValueObject::fromString($foo);
    }
    // ... process
}
```

While for the following I would suggest to split it up:

```php
function someHelper(SomeEntity|string|null $foo): ?string {
    if ($foo === null) {
        return null;
    }
    if (is_string($foo)) {
       $foo = $this->someRepository->findByFoo($foo);
    }
    // ... process
}
```

**Comment by markusguenther** [comment]
As @bwaidelich and @kitsunet talked about it. Do we want to adjust the helper like the proposal?

File: Neos.Media/Classes/Eel/AssetsHelper.php [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+        }
+        $collection = $this->assetCollectionRepository->findOneByTitle($collectionTitle);
+        return $this->findByCollection($collection);
+    }
+
+    /**
+     * @param Tag[] $tags
+     * @return QueryResultInterface<AssetInterface> | null
+     */
+    public function search(string $searchTerm, array $tags = [], AssetCollection $collection = null): ?QueryResultInterface
```
Comment (bwaidelich) [inline Neos.Media/Classes/Eel/AssetsHelper.php:103]:
```suggestion
    public function search(string $searchTerm, array $tags = [], AssetCollection $collection = null): ?QueryResultInterface
```

File: Neos.Media/Classes/Eel/AssetsHelper.php [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+    /**
+     * @Flow\Inject
+     * @var AssetCollectionRepository
+     */
+    protected $assetCollectionRepository;
+
+    /**
+     * @return QueryResultInterface<AssetInterface> | null
+     */
+    public function findByTag(?Tag $tag): ?QueryResultInterface
```
Comment (bwaidelich) [inline Neos.Media/Classes/Eel/AssetsHelper.php]: Maybe we _should_ allow polymorphism like Christian suggested:

```suggestion
    public function findByTag(Tag|string|null $tag): ?QueryResultInterface
```
?

File: Neos.Media/Classes/Eel/AssetsHelper.php [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+            return null;
+        }
+        $tag = $this->tagRepository->findOneByLabel($tagLabel);
+        return $this->findByTag($tag);
+    }
+
+    /**
+     * @return QueryResultInterface<AssetInterface> | null
+     */
+    public function findByCollection(?AssetCollection $collection): ?QueryResultInterface
```
Comment (bwaidelich) [inline Neos.Media/Classes/Eel/AssetsHelper.php]: And here!?
```suggestion
    public function findByCollection(AssetCollection|string|null $collection): ?QueryResultInterface
```

File: Neos.Media/Classes/Eel/AssetsHelper.php (hunk 2 above)
Comment (bwaidelich) [inline Neos.Media/Classes/Eel/AssetsHelper.php]: For this one it makes a lot of sense, too, IMO:

```suggestion
    /**
     * @param Tag[]|string[] $tags
     * @return QueryResultInterface<AssetInterface> | null
     */
    public function search(?string $searchTerm, array $tags = [], AssetCollection|string $collection = null): ?QueryResultInterface
```

**Review by bwaidelich — COMMENTED** [review]
> As @bwaidelich and @kitsunet talked about it.

We haven't really talked about it and I might have misinterpreted Christians point.

> Do we want to adjust the helper like the proposal?

Actually my suggestion was to keep it separate methods.. But thinking about it again: Tags are a special entity (and IMO they should have never been an entity but a value object)

So maybe it makes sense to adjust `findByTag()` to allow entity _or_ string (=label) like Christian suggested because you won't have an instance of the tag in many places.

And maybe the same for `findByCollection`.. I'm not sure

**Comment by Sebobo** [comment]
So I revert back to my first version that allowed both strings and objects for both methods?

**Comment by bwaidelich** [comment]
> So I revert back to my first version that allowed both strings and objects for both methods?

Sorry, I didn't follow this PR from the beginning and didn't know that this was the first version?
The suggestions just reflect my personal opinion and I'm not even sure about it – feel free to ignore me ;)

**Comment by bwaidelich** [comment]
OK, I think I finally understood Christians comment (I had read it looking at the _current_ version).

I'd say: Lets keep the separate methods as is.
But maybe we can change `search` as suggested? (i.e. allow term to be null and tags to be an array of strings)?

**Comment by mhsdesign** [comment]
I dindt completely follow the discussion, but can we later extract the conclusion into an issue or something, so we know how to write eel helpers in the future and dont have to re discuss this again ^^

**Comment by markusguenther** [comment]
The search does not work with tag labels, as the repository does not handle strings. If this is wanted, we need to transform all strings into tag instances first.

Also added some tests.

**Review by markusguenther — APPROVED** [review]

**Comment by markusguenther** [comment]
@bwaidelich Does this fit better?

**Comment by bwaidelich** [comment]
> Does this fit better?

IMO it's great like this, but this is just my personal preference and @kitsunet opposed this solution if I got it right?

> The search does not work with tag labels, as the repository does not handle strings.
> If this is wanted, we need to transform all strings into tag instances first.

That's what I would expect to be honest, like:

```php
$tagInstances = array_filter(array_map(fn (string|Tag $tag) => is_string($tag) ? $this->tagRepository->findOneByLabel($tag) : $tag));
```

**Comment by kitsunet** [comment]
It's fine by me, even though with the null (why did that come into play?) it now looks pretty convoluted with the type hints. I just think this will be hard to refactor later and fall on our feet sooner or later 😆 but if you all are fine with this, lets do it.

**Comment by bwaidelich** [comment]
> with the null (why did that come into play?)

I'm not sure, but IMO it's a good idea to be fault tolerant in Eel helpers, i.e. don't throw an exception for `${some.helper(foo)}` if `foo` happens to be `null`.
And many existing helpers already behave that way (either because they don't have type hints or because we explicitly allowed `null`, e.g. https://github.com/neos/neos-development-collection/blob/8.3/Neos.Fusion/Classes/Eel/BaseUriHelper.php#L31 or https://github.com/neos/neos-development-collection/blob/8.3/Neos.Neos/Classes/Fusion/Helper/NodeHelper.php#L56)

**Comment by Sebobo** [comment]
Yes. Often people feed variables into helpers which are not set yet, so we should be gentle.

**Comment by Sebobo** [comment]
@markusguenther I can make the tags adjustment in the search method later like we discussed. Thx for taking care of the PR while I was tripping.

**Comment by Sebobo** [comment]
I made the changes but now wonder, whether I shouldn't add a new search method to the AssetsRepository as search by tag OR collection is pretty useless IMO.

**Review by markusguenther — APPROVED** [review]

**Comment by markusguenther** [comment]
Wrote with @Sebobo yesterday, and we keep it as it is now.
So, the change of the `AssetRepository` can be a future improvement.

**Pull request merged**